<div class="alert alert-success">
<font size="5"><b>Комментарий ревьюера 2</b></font>

Привет еще раз. Спасибо, что доделал работу. Оформление комментариев по работе сохраняется. Только обозначим, что это вторая итерация. 

</div>

<div class="alert alert-success">
<font size="5"><b>Комментарий ревьюера</b></font>

Привет еще раз! Спасибо, что исправил задание:) Ты проделал большую работу над проектом. Он выполнен уже на достойном уровне. Однако есть моменты, которые еще можно улучшить. Будет здорово, если ты, надеюсь, не против, если я буду на ты, будешь отвечать на комментарии и участвовать в диалоге. Если обращение на ты неприемлемо, то прошу сообщить. 

Мои комментарии обозначены пометкой **Комментарий ревьюера**. Далее в файле ты сможешь найти их в похожих ячейках:
    
<div class="alert alert-success">Если фон комментария зелёный - всё сделано правильно. Рекомендации укажу таким же цветом;</div>
        
<div class="alert alert-warning">Оранжевый - некритичные замечания. Если таких замечаний меньше трех - проект может быть принят без их отработки;</div>
        
<div class="alert alert-danger">Красный - нужно переделать. </div>
        
Не удаляй эти комментарии и постарайся учесть их в ходе выполнения данного проекта. Свои же комментарии ты можешь обозначать любым заметным способом. 

</div>

<div class="alert alert-success">
<font size="5"><b>Комментарий ревьюера 0</b></font>

Привет, Руслан! Спасибо, что прислал задание:) Поздравляю с первым сданным проектом. Меня зовут Слепцов Артем и я буду проверять твой проект) К сожалению, я не могу запустить твой проект полностью из-за ошибки в коде. Посмотри, что пошло не так. По окончании работы перезапускай проект с целью проверки работоспособности кода. Как исправишь неточность, присылай проект снова. Буду ждать :)

</div>

**Привет, Артем! Да, на ты конечно просто и удобно)Рад знакомству! при первых правках затупил, не поздоровался в коде, пардон!**

<div class="alert alert-success">
<font size="5"><b>Комментарий ревьюера 2</b></font>

Бывает)

</div>

# Исследование надёжности заёмщиков

Заказчик — кредитный отдел банка. Нужно разобраться, влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок. Входные данные от банка — статистика о платёжеспособности клиентов.

Результаты исследования будут учтены при построении модели **кредитного скоринга** — специальной системы, которая оценивает способность потенциального заёмщика вернуть кредит банку.

<div class="alert alert-warning">
<font size="5"><b>Комментарий ревьюера</b></font>

Правильно, что есть краткое вступление в работу, описание того, что надо делать. В работе следует приводить информацию о входных данных: какие столбцы есть в таблице, их названия и какую информацию они несут. Так работа выглядит презентабельно. Подробнее смотри [методичку](https://www.notion.so/35d9c2f32ffa42f3b17010737aedca4b) по оформлению проектов. Она располагается в Notion в разделе "Как оформлять проект".

</div>

## Шаг 1. Откройте файл с данными и изучите общую информацию

In [1]:
import pandas as pd

Прочитаем файл data.csv и сохраним его в переменной df.

In [2]:
df = pd.read_csv('/datasets/data.csv')

Посмотрим данные в файле методом info

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       19351 non-null float64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        19351 non-null float64
purpose             21525 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


In [4]:
df.head(10)

children  days_employed  dob_years education  education_id  \
0         1   -8437.673028         42    высшее             0   
1         1   -4024.803754         36   среднее             1   
2         0   -5623.422610         33   Среднее             1   
3         3   -4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   
5         0    -926.185831         27    высшее             0   
6         0   -2879.202052         43    высшее             0   
7         0    -152.779569         50   СРЕДНЕЕ             1   
8         2   -6929.865299         35    ВЫСШЕЕ             0   
9         0   -2188.756445         41   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   
5  гражданский брак                 1      M   компаньон     0  255763.565419   
6   женат / замужем                 0      F   компаньон     0  240525.971920   
7   женат / замужем                 0      M   сотрудник     0  135823.934197   
8  гражданский брак                 1      F   сотрудник     0   95856.832424   
9   женат / замужем                 0      M   сотрудник     0  144425.938277   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу  
5               покупка жилья  
6           операции с жильем  
7                 образование  
8       на проведение свадьбы  
9     покупка жилья для семьи

In [5]:
df.duplicated().sum()

54

In [6]:
df.isnull().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Вывод**

<div class="alert alert-success">
<font size="5"><b>Комментарий ревьюера</b></font>

Подумай также о возможных причинах появления пропусков, а также о том, являются ли они случайными или нет. Можно высказать предварительные варианты их обработки. 

Первый взгляд на таблицу выполнен.

</div>

**Да, я с таким в текущей работе тоже сталкиваюсь постоянно. Пока еще я просто в гугл таблицах с такими вещами работаю, как провило там пропуски возниают когда изначально данных не получено. Например люди просто не хотят оставлять свой номер телефона, а поскольку шаг в лидформе не обязательный, то ячейка остается пустой. Возможно в этом случае такая же ситуация и была.**

<div class="alert alert-success">
<font size="5"><b>Комментарий ревьюера 2</b></font>

Да, вполне возможное предположение. 
</div>

## Шаг 2. Предобработка данных

### Обработка пропусков

Нам нужно проверить данные - правильность наименования столбцов, пропуски, дубликаты.

In [7]:
df.columns

Index(['children', 'days_employed', 'dob_years', 'education', 'education_id',
       'family_status', 'family_status_id', 'gender', 'income_type', 'debt',
       'total_income', 'purpose'],
      dtype='object')

В названиях столбцов ошибок стилистики нет, все написано в змеином регистре.

Посмотрим на данные отдельно в стобцах 'days_employed' и 'total_income'

In [8]:
df['days_employed'].head(50)

0      -8437.673028
1      -4024.803754
2      -5623.422610
3      -4124.747207
4     340266.072047
5       -926.185831
6      -2879.202052
7       -152.779569
8      -6929.865299
9      -2188.756445
10     -4171.483647
11      -792.701887
12              NaN
13     -1846.641941
14     -1844.956182
15      -972.364419
16     -1719.934226
17     -2369.999720
18    400281.136913
19    -10038.818549
20     -1311.604166
21      -253.685166
22     -1766.644138
23      -272.981385
24    338551.952911
25    363548.489348
26              NaN
27      -529.191635
28      -717.274324
29              NaN
30    335581.668515
31     -1682.083438
32     -4649.910832
33     -1548.637544
34     -4488.067031
35    394021.072184
36      -176.216688
37     -6448.810860
38      -597.881827
39      -650.587796
40     -1030.362606
41              NaN
42     -1257.496190
43     -4375.681384
44     -1362.041728
45     -1039.451741
46     -2262.712304
47     -2689.137274
48     -3341.067886
49     -1181.443228


В столбце общего трудового стажа в днях есть пропуски, при этом значения записаны и в отрицательном выражении и с дробной частью. Пропуски могут возникнуть в данном случае либо из-за отсутствия информации, либо из-за неверного формата записи.

In [9]:
df['total_income'].head(50)

0     253875.639453
1     112080.014102
2     145885.952297
3     267628.550329
4     158616.077870
5     255763.565419
6     240525.971920
7     135823.934197
8      95856.832424
9     144425.938277
10    113943.491460
11     77069.234271
12              NaN
13    130458.228857
14    165127.911772
15    116820.904450
16    289202.704229
17     90410.586745
18     56823.777243
19    242831.107982
20    209552.644578
21    131812.203909
22    149681.140731
23    128265.720871
24    290547.235997
25     55112.757732
26              NaN
27    308848.983691
28    187863.237306
29              NaN
30    171456.067993
31    275485.684538
32    139057.464207
33    157245.786233
34    104660.449203
35     77805.677436
36    138830.650734
37    154128.878672
38    192247.298639
39    754240.797248
40    142865.579146
41              NaN
42     82065.089021
43    272957.133509
44    100780.496749
45    106385.919724
46    182682.459898
47    356277.909345
48    162065.516450
49     44591.806888


Аналогично, в столбце ежемесячного дохода присутствуют пропуски, вероятно, по тем же причинам. Нужно понять какими значениями мы можем заменить пропуски - напрмиер нулями или средними/медианными значениями по столбцу или вообще их удалить.  
Поэтому проверим есть ли взаимосвязь между пропусками в этих двух столбцах (в одинаковых ли строках пропуски)


In [10]:
df[df['total_income'].isna()].head(10)

children  days_employed  dob_years education  education_id  \
12         0            NaN         65   среднее             1   
26         0            NaN         41   среднее             1   
29         0            NaN         63   среднее             1   
41         0            NaN         50   среднее             1   
55         0            NaN         54   среднее             1   
65         0            NaN         21   среднее             1   
67         0            NaN         52    высшее             0   
72         1            NaN         32    высшее             0   
82         2            NaN         50    высшее             0   
83         0            NaN         52   среднее             1   

            family_status  family_status_id gender  income_type  debt  \
12       гражданский брак                 1      M    пенсионер     0   
26        женат / замужем                 0      M  госслужащий     0   
29  Не женат / не замужем                 4      F    пенсионер     0   
41        женат / замужем                 0      F  госслужащий     0   
55       гражданский брак                 1      F    пенсионер     1   
65  Не женат / не замужем                 4      M    компаньон     0   
67        женат / замужем                 0      F    пенсионер     0   
72        женат / замужем                 0      M  госслужащий     0   
82        женат / замужем                 0      F    сотрудник     0   
83        женат / замужем                 0      M    сотрудник     0   

    total_income                                purpose  
12           NaN                        сыграть свадьбу  
26           NaN                            образование  
29           NaN       строительство жилой недвижимости  
41           NaN       сделка с подержанным автомобилем  
55           NaN                        сыграть свадьбу  
65           NaN  операции с коммерческой недвижимостью  
67           NaN                покупка жилья для семьи  
72           NaN  операции с коммерческой недвижимостью  
82           NaN                                  жилье  
83           NaN                                  жилье

Пропуски совпадают по строкам. При этом на примере первых 10 строк мы видим, что данные о стаже и доходе пропущены у людей в возрасте, при этом отмечен типа занятости (напрмиер госслужащий точно имеет стаж и доход). Скорее всего значения days_employed и total_income были пропущены или потеряны. Заменить их нулями будет некорректно, поскольку это может повлиять на результат. Строк с пропусками 10,1% поэтому удалять их тоже нельзя.

Для замены пропусков в этих столбцах нужно использовать методы средних чисел и медианы. Средние числа подойдут для столбца df['days_employed'], а медиана для df['total_income'], потому что при расчете зарплат медиана исключает аномалии и отображает более реальные цифры, чем среднее арифмитическое. 

Для того, чтобы без повреждения данных заполнить пропуски, воспользуемся изучением данных по группам и впишем значения в рамках группы. Проводить группировку будем по колонк income_type (тип занятости), поскольку профессия кореллирует с зарплатой и стажем.

In [11]:
pd.pivot_table(df,
               index=['income_type'],
              values=['days_employed'])

days_employed
income_type                   
безработный      366413.652744
в декрете         -3296.759962
госслужащий       -3399.896902
компаньон         -2111.524398
пенсионер        365003.491245
предприниматель    -520.848083
сотрудник         -2326.499216
студент            -578.751554

In [12]:
df['income_type'][df['days_employed'] >= 0].unique()
df['income_type'][df['days_employed'] >= 0].value_counts()

пенсионер      3443
безработный       2
Name: income_type, dtype: int64

In [13]:
df['income_type'][df['days_employed'] < 0].unique()
df['income_type'][df['days_employed'] < 0].value_counts()

сотрудник          10014
компаньон           4577
госслужащий         1312
предприниматель        1
в декрете              1
студент                1
Name: income_type, dtype: int64

То есть 3443 пенсионера и 2 безработных с положительными значениями - это первая группа. Остальные данные по типу занятости с отрицательными значениями попадают во вторую группу. Теперь в группах нам нужно заменить пропуски на средние или медианные для группы значения. Для этого создадим переменную, хранящую среднее значение по группировке и воспользуемся методами isin и isna - для поиска пропущенных значений в группах и замены на новую переменную. 

In [14]:
first_group_days_employed_mean = df['days_employed'][df['days_employed'] > 0].mean()
df['days_employed'][(df['income_type'].isin(['пенсионер', 'безработный'])) & (df['days_employed'].isna())] = first_group_days_employed_mean

secong_group_days_employed_mean = df['days_employed'][df['days_employed'] <= 0].mean()
df['days_employed'][(df['income_type'].isin(['сотрудник', 'компаньон', 'госслужащий', 'в декрете', 'студент', 'предприниматель'])) & (df['days_employed'].isna())] = secong_group_days_employed_mean
pd.set_option('chained_assignment', None)

/opt/conda/lib/python3.7/site-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  
/opt/conda/lib/python3.7/site-packages/ipykernel_launcher.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """


<div class="alert alert-success">
<font size="5"><b>Комментарий ревьюера 2</b></font>

Попробуй использовать команду pd.set_option('chained_assignment', None) для скрытия данного предупреждения.

</div>

Сделаем то же самое для total_income, но используем метод медианы

In [15]:
first_group_income_median = df['total_income'][df['days_employed'] > 0].median()
df['total_income'][(df['income_type'].isin(['пенсионер', 'безработный'])) & (df['total_income'].isna())] = first_group_income_median

second_group_income_median = df['total_income'][df['days_employed'] <= 0].median()
df['total_income'][(df['income_type'].isin(['сотрудник', 'компаньон', 'госслужащий', 'в декрете', 'студент', 'предприниматель'])) & (df['total_income'].isna())] = second_group_income_median
pd.set_option('chained_assignment', None)

Далее нужно проверить, что пропущенных значений больше нет. Снова подсчитаем сумму пропусков

In [16]:
df.isnull().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

**Вывод**

Были пропущены значения в столбцах days_employed и total_income. Для замены пропущенных значений использовали методы средней арифметической и медианы, чтобы не потерять часть данных и приблизить их к корректным. После замены значений пропуски отсутствуют. Теперь можем приступить к рассмотрению типов данных.

<div class="alert alert-danger">
<font size="5"><b>Комментарий ревьюера</b></font>

Обработать пропуски просто медианой или средним недостаточно. Лучше использовать заполнение пропусков на основе группировки данных по одному из параметров в нашей таблице. Подумай, какой параметр будет больше всего влиять на доход. От чего больше всего зависит доход? Проведи группировку по нему и на его основании заполни пропуски. 

</div>

**Вероятно, доход зависит от типа занятости (колонка income_type). Я попробовал сначала сделать две переменные, в первую сохранить среднее значение по группе, во вторую сохранить список индесов пропущенных значений и затем заменить вторые на первые. Но что-то пошло не так, не получалось. Использовал isin, все получилось. Блокнот ругается на меня, выдавая сообщение, что "Значение пытается быть установлено для копии среза из фрейма данных" - я не совсем понял почему, помоги понять? Может надо было использовать loc?**

<div class="alert alert-success">
<font size="5"><b>Комментарий ревьюера 2</b></font>

Да, лучше работать через loc. Но предупреждение - этоо не так страшно. Выше привел способ его скртыия. 

</div>

<div class="alert alert-warning">
<font size="5"><b>Комментарий ревьюера 2</b></font>

Но вообще способ не самый лучший, поскольку пропуски ты заполняешь одним значением для нескольких групп. Согласись, доходы студентов и госслужащих все же будут отличаться. Поэтому стоило взять цикл по уникальным типам занятости, в нем рассчитывать медианной доход для каждой группы и затем заполнять пропуски в данном типе занятости рассчитанным значением при помощи метода fillna(). 

</div>

### Замена типа данных

Ранее мы провели предобработку пропущенных значений в столбцах days_employed и total_income. Эти же 2 столбца являются типом float64. Для нашег оисследования нам не нужна дробная часть или отрицательные значеняи в этих величинах, поэтому для удобного сравнения проведем замену типа данных с вещественного на целочисленные значения. Применим метод try/except , чтобы увидеть возможные ошибки в ходе выполнения операции.

In [17]:
try:
    df['days_employed'] = df['days_employed'].round().astype('int64')
    df['total_income'] = df['total_income'].round().astype('int64')
except:
    'Ошибка'

In [18]:
df.head()

children  days_employed  dob_years education  education_id  \
0         1          -8438         42    высшее             0   
1         1          -4025         36   среднее             1   
2         0          -5623         33   Среднее             1   
3         3          -4125         32   среднее             1   
4         0         340266         53   среднее             1   

      family_status  family_status_id gender income_type  debt  total_income  \
0   женат / замужем                 0      F   сотрудник     0        253876   
1   женат / замужем                 0      F   сотрудник     0        112080   
2   женат / замужем                 0      M   сотрудник     0        145886   
3   женат / замужем                 0      M   сотрудник     0        267629   
4  гражданский брак                 1      F   пенсионер     0        158616   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу

В таблице остались только верные значения. Теперь можно применить функцию abs для замены отрицательных значений на положительные в столбце days_employed.

In [19]:
df['days_employed'] = df['days_employed'].abs()

Проверим вывод

In [20]:
df.head()

children  days_employed  dob_years education  education_id  \
0         1           8438         42    высшее             0   
1         1           4025         36   среднее             1   
2         0           5623         33   Среднее             1   
3         3           4125         32   среднее             1   
4         0         340266         53   среднее             1   

      family_status  family_status_id gender income_type  debt  total_income  \
0   женат / замужем                 0      F   сотрудник     0        253876   
1   женат / замужем                 0      F   сотрудник     0        112080   
2   женат / замужем                 0      M   сотрудник     0        145886   
3   женат / замужем                 0      M   сотрудник     0        267629   
4  гражданский брак                 1      F   пенсионер     0        158616   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу

Проверим тип данных после всех операций

In [21]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       21525 non-null int64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        21525 non-null int64
purpose             21525 non-null object
dtypes: int64(7), object(5)
memory usage: 2.0+ MB


В исходных данных остались только целочисленные и строчные значения.

**Вывод**

Для удобной работы с данными в рамках этого исследования мы провели операции замены типа данных на целочисленные значения и избавились от отрицательных значений. Теперь в таблице остались только строковые и целочисленные значения и мы можем приступит ьк обработке дубликатов.

<div class="alert alert-success">
<font size="5"><b>Комментарий ревьюера</b></font>

Исходя из общей информации о таблице мы видим, что два столбца имеют вещественный тип данных, который и надо заменить на целочисленный. Эта замена произведена верно. Также хорошо бы освоить метод to_numeric() для будущих работ. 

</div>

**Спасибо! Я не знаю почему, но какие то медотды прям хорошо в глову ложатся, а какие то вот убей не могу без подсказки выполнить. Дальеш будет жетсче, поэтому шпаргалки наше все))**

### Обработка дубликатов

В начале исследования мы нашли 54 дубликата. Удалим полные дубликаты методом drop_duplicates() и перезапишем 

In [22]:
df = df.drop_duplicates().reset_index(drop = True)

Проверим сколько осталось строк в датасете

In [23]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21471 entries, 0 to 21470
Data columns (total 12 columns):
children            21471 non-null int64
days_employed       21471 non-null int64
dob_years           21471 non-null int64
education           21471 non-null object
education_id        21471 non-null int64
family_status       21471 non-null object
family_status_id    21471 non-null int64
gender              21471 non-null object
income_type         21471 non-null object
debt                21471 non-null int64
total_income        21471 non-null int64
purpose             21471 non-null object
dtypes: int64(7), object(5)
memory usage: 2.0+ MB


В датасете осталось 21471 значение, а было 21525, то есть удаление 54 дубликатов прошло успешно. Однако на первом шаге при изучении информации мы обнаружили, что в столбце education разное написание уровня образования. Проверим сколько враиантов написания и каких методом value_counts в столбце

In [24]:
df['education'].value_counts()

среднее                13705
высшее                  4710
СРЕДНЕЕ                  772
Среднее                  711
неоконченное высшее      668
ВЫСШЕЕ                   273
Высшее                   268
начальное                250
Неоконченное высшее       47
НЕОКОНЧЕННОЕ ВЫСШЕЕ       29
НАЧАЛЬНОЕ                 17
Начальное                 15
ученая степень             4
Ученая степень             1
УЧЕНАЯ СТЕПЕНЬ             1
Name: education, dtype: int64

<div class="alert alert-success">
<font size="5"><b>Комментарий ревьюера</b></font>

Здорово, что ты заметил данную неточность в данных.

</div>

Видим, что орфографическое написание одинаковое, а регистры разные. Это может усложнить работу с документом и повлиять на результаты, поэтому приведем все значения к нижнему регистру

In [25]:
df['education'] = df['education'].str.lower()
df['education'].value_counts()

среднее                15188
высшее                  5251
неоконченное высшее      744
начальное                282
ученая степень             6
Name: education, dtype: int64

После того, как мы изменили данные, нужно провести повторную проверку дубликатов, возможно появятся еще строки, полностью идентичные.

In [26]:
df.duplicated().sum()

17

Действительно, мы нашли таким образом еще 17 полных дубликатов. Удалим их из таблицы.

In [27]:
df = df.drop_duplicates().reset_index(drop = True)

In [28]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21454 entries, 0 to 21453
Data columns (total 12 columns):
children            21454 non-null int64
days_employed       21454 non-null int64
dob_years           21454 non-null int64
education           21454 non-null object
education_id        21454 non-null int64
family_status       21454 non-null object
family_status_id    21454 non-null int64
gender              21454 non-null object
income_type         21454 non-null object
debt                21454 non-null int64
total_income        21454 non-null int64
purpose             21454 non-null object
dtypes: int64(7), object(5)
memory usage: 2.0+ MB


Вероятно ,если проверить другие столбцы на уникальные значения, мы также сможем удалить некоторые дубликаты.

In [29]:
df['family_status'].unique()

array(['женат / замужем', 'гражданский брак', 'вдовец / вдова',
       'в разводе', 'Не женат / не замужем'], dtype=object)

In [30]:
df['gender'].unique()

array(['F', 'M', 'XNA'], dtype=object)

In [31]:
df['income_type'].unique()

array(['сотрудник', 'пенсионер', 'компаньон', 'госслужащий',
       'безработный', 'предприниматель', 'студент', 'в декрете'],
      dtype=object)

In [32]:
df['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

В колонке family_status есть значение 'Не женат / не замужем' - приведем к нижнему регистру.

In [33]:
df['family_status'] = df['family_status'].str.lower()

В колонке purpose встречается много похожих значений, например 'сыграть свадьбу' и 'на проведение свадьбы'. Это одна и та же цель получения средств, но записана разным способом. Для того, чтобы нам объединить одинаковые цели необходимо провести лемматизацию значений 

**Вывод**

Мы нашли и удалили из таблицы 71 строку дубликатов в столбце 'education'. Мы обнаружили, что в столбце purpose есть схожие цели, н оразнео написание, поэтому нужно провести лемматизацию значений.

<div class="alert alert-success">
<font size="5"><b>Комментарий ревьюера</b></font>

Дубликаты вызывают смещение финальных результатов, поэтому их и надо удалять. Конечно, в данных нет уникального идентификатора, однако надо уметь удалять дубликаты. Метод drop_duplicates() использован верно.

</div>

### Лемматизация

Сначала нам надо вызвать метод pymystem3 и провреить все уникальные значения из столбца purpose

In [34]:
from pymystem3 import Mystem
from collections import Counter
m = Mystem()
word_list = []
for value in df['purpose'].unique():
    lemmas = m.lemmatize(value)
    word_list += lemmas

print(Counter(word_list))

Counter({' ': 59, '\n': 38, 'покупка': 10, 'недвижимость': 10, 'автомобиль': 9, 'образование': 9, 'жилье': 7, 'с': 5, 'операция': 4, 'на': 4, 'свой': 4, 'свадьба': 3, 'строительство': 3, 'получение': 3, 'высокий': 3, 'дополнительный': 2, 'для': 2, 'коммерческий': 2, 'жилой': 2, 'заниматься': 2, 'сделка': 2, 'приобретение': 1, 'сыграть': 1, 'проведение': 1, 'семья': 1, 'собственный': 1, 'подержать': 1, 'со': 1, 'подержанный': 1, 'профильный': 1, 'сдача': 1, 'ремонт': 1})


Создадим список из слов по целям кредитования, по которым проведем лемматизацию.

In [35]:
purpose_list = ['свадьба','автомобиль','недвижимость','образование','жилье']

In [36]:
def purpose_choose(row):
    purpose = row['purpose']
    lemmas = m.lemmatize(purpose)
    for purpose_type in purpose_list:
        if (purpose_type in lemmas):
            return purpose_type



df['purpose_type'] = df.apply(purpose_choose, axis=1) 
df['purpose_type'].value_counts()

недвижимость    6351
жилье           4460
автомобиль      4306
образование     4013
свадьба         2324
Name: purpose_type, dtype: int64

Теперь надо проверить наличие дубликатов в датасете, но исключив изначальный столбец purpose, удалить их и проверить количество строк методом info


In [37]:
df = df.drop(['purpose'], axis = 1)
df.duplicated().sum()

244

In [38]:
df = df.drop_duplicates().reset_index(drop = True)
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21210 entries, 0 to 21209
Data columns (total 12 columns):
children            21210 non-null int64
days_employed       21210 non-null int64
dob_years           21210 non-null int64
education           21210 non-null object
education_id        21210 non-null int64
family_status       21210 non-null object
family_status_id    21210 non-null int64
gender              21210 non-null object
income_type         21210 non-null object
debt                21210 non-null int64
total_income        21210 non-null int64
purpose_type        21210 non-null object
dtypes: int64(7), object(5)
memory usage: 1.9+ MB


У нас получилось 21210 строк, размер файла уменьшился на 0,1 МВ

**Вывод**

Лемматизация помогла нам собрать все значения столбца purpose из исходных данных в 5 сегментов. Мы удалили еще 244 строки дубликатов.
Природа возникновения дубликатов неизвестно, но разный формат записей значений говорит либо об отсутствии единого стандарта заполнения данных в анкете, либо о свободной форме заполнения. В любом случае теперь мы можем категоризировать данные и провести нужный анализ в рамках исследования.

<div class="alert alert-success">
<font size="5"><b>Комментарий ревьюера</b></font>

Круто, ты сделал самую сложную часть проекта. Использовал нужные методы, и даже провел группировку с помощью собственной функции. Выбор основных целей кредитов обоснован. 

</div>

**С собвственными функциями это самая большая моя проблема, поэтому т тя откровенно просил помощи в объяснении у товарищей. Писал сам но попыток было столько, что мама не горюй. Надеюсь дальше мне будет проще работать со своими функциями, ибо удобно очень.**

<div class="alert alert-success">
<font size="5"><b>Комментарий ревьюера 2</b></font>

Но стоит сказать, что функция получилась классной!

</div>

### Категоризация данных

Проведем категоризацию данных для првоедения исследования. Изучим количество детей и записей, а также данные по группировке.

In [39]:
df['children'].value_counts()

 0     13892
 1      4773
 2      2043
 3       329
 20       76
-1        47
 4        41
 5         9
Name: children, dtype: int64

В данных присутствует сегмент с показателем -1. Скорее всего это ошибка данных или ввода. Количество записей всего 47, для проведения исследования можо присвоить эти данные к самой большой категории - 0 детей. (Хотя на самом деле лучше обратиться к менеджерам и уточнить про этот параметр, почему он оказался в файле как -1, возможно здесь должен быть просто 1 ребенок).

In [40]:
df.groupby('children').mean()

days_employed  dob_years  education_id  family_status_id      debt  \
children                                                                       
-1         63171.510638  42.574468      0.829787          0.957447  0.021277   
 0         92261.126836  46.174201      0.829758          1.130507  0.076447   
 1         22896.923109  38.375864      0.786298          0.812696  0.093023   
 2          5731.701419  35.764562      0.790015          0.452766  0.094958   
 3          8595.267477  36.264438      0.823708          0.404255  0.082067   
 4         11897.829268  36.048780      0.780488          0.512195  0.097561   
 5          1534.555556  38.777778      1.222222          0.222222  0.000000   
 20        44753.421053  41.815789      0.842105          0.815789  0.105263   

           total_income  
children                 
-1        152995.000000  
 0        163281.528002  
 1        169267.536769  
 2        169857.335781  
 3        179848.167173  
 4        167244.731707  
 5        168286.222222  
 20       165779.657895

Если сравнить данные по количеству детей, то в целом они примерно одинаковые, однако в колонке debt у сегмента с -1 ребенком самый низкий показатель. 

In [41]:
df['children'] = df['children'].replace(-1, 0)

Для удобства анализа разделим данные на 4 категории -  0, 1, 2 и более детей. Создадим новый столбец 'children_category'

In [42]:
def children_category(children):
        if children == 0:
                return 0
        if children == 1:
                return 1
        if children == 2:
                return 2
        return 3
    
df['children_category'] = df['children'].apply(children_category)
df.groupby('children_category')['children'].count()

children_category
0    13939
1     4773
2     2043
3      455
Name: children, dtype: int64

<div class="alert alert-success">
<font size="5"><b>Комментарий ревьюера</b></font>

Лучше тогда назвать группу 3+.

</div>

Нам нужно выяснить влияние количества детей и уровня дохода на возврат кредита в срок, поэтому проведем категоризацию и по доходу и создадим новый столбец income_level.

In [43]:
def income_category(income):

        if income < 100000:
                return 'низкий доход'
        if income <= 200000:
                return 'средний доход'
        return 'высокий доход'
    
df['income_level'] = df['total_income'].apply(income_category)
df.groupby('income_level')['children'].count()

income_level
высокий доход     5066
низкий доход      4463
средний доход    11681
Name: children, dtype: int64

<div class="alert alert-warning">
<font size="5"><b>Комментарий ревьюера</b></font>

Исходя из чего выбран именно такой способ категоризации столбца с доходами? Такое разбиение мы можем использовать, но важно его обосновать. 
    
</div>
<div class="alert alert-success">
    
Лучше использовать разбиение по квантилям данных (метод qcut). Тогда группы получатся одинакового размера. Или же можно узнать в интернете процентный состав общества по доходам и затем уже категоризовать столбец с доходами по найденному соотношению с помощью персентилей. 

</div>

**Здесь логика у меня была такая - самая наиболее часто встречающаяся во всех отчетах категоризация дохода это три уровня - высокий, средний и низкий. Цифры с ноликами взяты для четкого деления дохода, как рубежи. В рамках исследовани яне стояло цели создать свою систему ранжирования уровня дохода, поэтому категоризировал "как везде".**

<div class="alert alert-success">
<font size="5"><b>Комментарий ревьюера 2</b></font>

Ок. Но лучше все же привести источник информации с указанием подобного способа категоризации. Ссылки - всегда хорошо)

</div>

In [44]:
df.head()

children  days_employed  dob_years education  education_id  \
0         1           8438         42    высшее             0   
1         1           4025         36   среднее             1   
2         0           5623         33   среднее             1   
3         3           4125         32   среднее             1   
4         0         340266         53   среднее             1   

      family_status  family_status_id gender income_type  debt  total_income  \
0   женат / замужем                 0      F   сотрудник     0        253876   
1   женат / замужем                 0      F   сотрудник     0        112080   
2   женат / замужем                 0      M   сотрудник     0        145886   
3   женат / замужем                 0      M   сотрудник     0        267629   
4  гражданский брак                 1      F   пенсионер     0        158616   

  purpose_type  children_category   income_level  
0        жилье                  1  высокий доход  
1   автомобиль                  1  средний доход  
2        жилье                  0  средний доход  
3  образование                  3  высокий доход  
4      свадьба                  0  средний доход

Для оптимизации датасета выделим словари с категориями образования и семейного положения

In [45]:
df_log = df[['days_employed', 'dob_years', 'education_id',
       'family_status_id', 'gender', 'income_type', 'debt',
       'purpose_type', 'children_category', 'income_level']]

df_family_dict = df[['family_status', 'family_status_id']]
df_family_dict = df_family_dict.drop_duplicates().reset_index(drop = True)

df_education_dict = df[['education', 'education_id']]
df_education_dict = df_education_dict.drop_duplicates().reset_index(drop = True)

**Вывод**

Для классификации датасета по параметрам с числовыми значениями мы разделили значения столбца children на четыре категории и столбца total_income на три категории.
Проведя категоризацию данных по количеству детей и доходу мы видим что по количеству детей самая большая категория - 0 детей, а по уровню дохода самая большая категория - средний.

<div class="alert alert-success">
<font size="5"><b>Комментарий ревьюера</b></font>

Словари в данных выделены. Правильно, что ты убрал текстовые параметры из основной таблицы. Это позволит нам увеличить эффективность работы с данными. 

</div>

## Шаг 3. Ответьте на вопросы

- Есть ли зависимость между наличием детей и возвратом кредита в срок?

In [46]:
pd.pivot_table(df,
               index=['children_category'],
              values=['debt'])

debt
children_category          
0                  0.076261
1                  0.093023
2                  0.094958
3                  0.085714

Процент клиентов с детьми, имеющих задолженности : 9.31%, 9.49%, 8.57%, а у тех, кого нет детей 7.62%

**Вывод**

Самый низкий процент должников среди категории без детей - у 7,62% людей. А у тех кого есть дети показатели выше  -  9,31%. Интересно что показатель процента задолжности у группы с 1 или 2 ребенком примерно одинаковый. Вероятно дети влияют на выплату в срок.

<div class="alert alert-success">
<font size="5"><b>Комментарий ревьюера</b></font>

Вывод не противоречит полученному результату. 

</div>

- Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [47]:
pd.pivot_table(df,
               index=['family_status'],
              values=['debt'])

debt
family_status                  
в разводе              0.071249
вдовец / вдова         0.066456
гражданский брак       0.093969
женат / замужем        0.076556
не женат / не замужем  0.098138

Посмотрим средний возраст по категориям, чтобы выстроить гипотезу и дать ответ на вопрос.

In [48]:
df_family_age_grouped = df.groupby('family_status')['dob_years'].mean().sort_values(ascending = True)
print('Средний возраст по категориям ', df_family_age_grouped)

Средний возраст по категориям  family_status
не женат / не замужем    38.380014
гражданский брак         42.024946
женат / замужем          43.476457
в разводе                45.503772
вдовец / вдова           56.443038
Name: dob_years, dtype: float64


**Вывод**

Больше всего должников среди не женатых, гед средний возраст составляет 38 лет. Более возрастные категории я вляются более стабильными плательщиками, особенно категория вдовцы - всего 6.6% должников, средний возраст 56 лет. Зависимост ьмежду семейным положением и возвратом кредита в срок есть.

<div class="alert alert-success">
<font size="5"><b>Комментарий ревьюера</b></font>

Скоррее всего мы имеем скрытую зависимость от возраста клиентов. 

</div>

**Что такое скрытая зависимость? Я так понимю, это когда все переменные уравнения неизвестны, но по известной части мы делаем правильный вывод? Например, была бы у нас информация о том, снимает человек жилье или живет в собственном, мы могли бы провести параллель и выяснить что молодые чаще снимают жилье, поэтому чаще остаются без денег на момент выплаты кредита, как то так?**

<div class="alert alert-success">
<font size="5"><b>Комментарий ревьюера 2</b></font>

В данном случае просто фактор отсутствия штампа сам по себе не несет информации о том, что клиент с большей вероятностью окажется должником. Выше мы смотрели на число детей, там фактор детей подразумевает дополнительные траты, а тут нет. Соответственно, есть еще какая-то зависимость, которая и формирует такой результат, который мы видим мы. Ты ее обнаружил правильно, на мой взгляд. Возраст клиентов оказывает большее влияние, чем семейное положение.  

</div>

- Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [49]:
pd.pivot_table(df,
               index=['income_level'],
              values=['debt'])

debt
income_level           
высокий доход  0.070667
низкий доход   0.079319
средний доход  0.088006

**Вывод**

Никакой ярко выраженной прямой зависимости между уровнем дохода и невыплатами кредита нет.Чуть выше процент должников со средним доходом, но это не выглядит сильными разрывом.

- Как разные цели кредита влияют на его возврат в срок?

In [50]:
pd.pivot_table(df,
               index=['purpose_type'],
              values=['debt'])

debt
purpose_type          
автомобиль    0.094101
жилье         0.069873
недвижимость  0.075719
образование   0.093340
свадьба       0.080659

**Вывод**

Лидирующие категории по количеству должников - те кто занимают на автомобиль или образование. Вероятно эти кредиты проще получить, чем кредит на недвижимость или жилье. Также достаточно высокий процент должников в категории свадеб.

<div class="alert alert-success">
<font size="5"><b>Комментарий ревьюера</b></font>

Здорово, что при анализе использован метод сводных таблиц. Он пригодится в дальнейших проектах.

Выводы сделаны по всем пунктам. 

</div>

## Шаг 4. Общий вывод

Мы провели классификацию данных о клиентах банков и определили ряд закономерностей в поведении разных групп людей.

Клиенты без детей реже задерживают платежи.

Семейный статус клиента влияет на вероятность возврата им кредита в срок.

По указанной цели кредита можно определить риск невозврата.

Уровень дохода клиента не определяет его пунктуальность в платежах.

<div class="alert alert-success">
<font size="5"><b>Комментарий ревьюера</b></font>

Выводы сделаны. Приведены ответы на главные вопросы проекта. В выводах можно приводить полученные ранее значения, верно. Также можно расписать все, что было сделано в ходе проведения работы. 

</div>

<div class="alert alert-success">
<font size="5"><b>Комментарий ревьюера</b></font>

Ты проделал большую работу, молодец! У тебя получился классный проект) Критических замечаний почти нет. Попробуй улучшить обработку пропусков. Думаю, сделать это тебе не составит труда. Жду твою работу :)

</div>

**Спасибо большое! Я дико отстаю от плана спринтов, уже надо за сборный проект браться, но мне еще одну работу предстоит выполнить перед этим. Надеюсь снова попасть к тебе на првоерку!**

<div class="alert alert-success">
<font size="5"><b>Комментарий ревьюера 2</b></font>

Помарки исправлены, и теперь работа выполнена хорошо. У тебя получился классный проект, молодец. Продолжай в том же духе! Надеюсь, ты успеешь в сроки) Успехов в дальнейших проектах :)

</div>
   

## Чек-лист готовности проекта

Поставьте 'x' в выполненных пунктах. Далее нажмите Shift+Enter.

- [x]  открыт файл;
- [х]  файл изучен;
- [х]  определены пропущенные значения;
- [х]  заполнены пропущенные значения;
- [х]  есть пояснение, какие пропущенные значения обнаружены;
- [х]  описаны возможные причины появления пропусков в данных;
- [х]  объяснено, по какому принципу заполнены пропуски;
- [х ]  заменен вещественный тип данных на целочисленный;
- [ х]  есть пояснение, какой метод используется для изменения типа данных и почему;
- [ х]  удалены дубликаты;
- [ х]  есть пояснение, какой метод используется для поиска и удаления дубликатов;
- [ х]  описаны возможные причины появления дубликатов в данных;
- [х ]  выделены леммы в значениях столбца с целями получения кредита;
- [х ]  описан процесс лемматизации;
- [х ]  данные категоризированы;
- [х ]  есть объяснение принципа категоризации данных;
- [ х]  есть ответ на вопрос: "Есть ли зависимость между наличием детей и возвратом кредита в срок?";
- [х ]  есть ответ на вопрос: "Есть ли зависимость между семейным положением и возвратом кредита в срок?";
- [ х]  есть ответ на вопрос: "Есть ли зависимость между уровнем дохода и возвратом кредита в срок?";
- [х]  есть ответ на вопрос: "Как разные цели кредита влияют на его возврат в срок?";
- [ х]  в каждом этапе есть выводы;
- [х ]  есть общий вывод.